### PR TITLE
adding documentation for hpa custom metrics

### PR DESCRIPTION
**Pull request by enoodle**
This documents changes made in : https://github.com/knative/serving/pull/12277

## Proposed Changes 

- Add documentation for HPA autoscaling by custom metrics

### CODE REVIEW

File: docs/serving/autoscaling/autoscaling-metrics.md [hunk 1]
```diff
@@ -15,52 +15,89 @@ The possible metric types that can be configured per revision depend on the type
...
+  template:
+    metadata:
+      annotations:
+        autoscaling.knative.dev/class: "hpa.autoscaling.knative.dev"
+        autoscaling.knative.dev/metric: "memory"
+```
+
+=== Per-revision custom metric configuration
+Assuming that you have a custom metric named `metricName` you can create an HPA to autoscale the revision by that metric.
+The HPA will be configured to use the average value of that metric over all the pods of the revision.
```
Comment (nader-ziada) [inline docs/serving/autoscaling/autoscaling-metrics.md:86]:
```suggestion
The HPA will be configured to use the **average value** of that metric over all the pods of the revision.
```

File: docs/serving/autoscaling/autoscaling-metrics.md [hunk 2]
```diff
@@ -15,52 +15,89 @@ The possible metric types that can be configured per revision depend on the type
 For more information about KPA and HPA, see the documentation on [Supported Autoscaler types](autoscaler-types.md).
 
 * **Per-revision annotation key:** `autoscaling.knative.dev/metric`
-* **Possible values:** `"concurrency"`, `"rps"` or `"cpu"`, depending on your Autoscaler type. The `cpu` metric is only supported on revisions with the HPA class.
+* **Possible values:** `"concurrency"`, `"rps"`, `"cpu"`, `"memory"` or any custom metric name, depending on your Autoscaler type. The `cpu`/`memory`/custom metrics are only supported on revisions with the HPA class.
```
Comment (nader-ziada) [inline docs/serving/autoscaling/autoscaling-metrics.md:18]:
```suggestion
* **Possible values:** `"concurrency"`, `"rps"`, `"cpu"`, `"memory"` or any custom metric name, depending on your Autoscaler type. The `cpu`/`memory`/`custom` metrics are only supported on revisions with the HPA class.
```

**Review by nader-ziada — COMMENTED** [review]
thanks @enoodle for the docs updates

File: docs/serving/autoscaling/autoscaling-metrics.md (hunk 2 above)
Comment (enoodle) [reply docs/serving/autoscaling/autoscaling-metrics.md]: :+1:

File: docs/serving/autoscaling/autoscaling-metrics.md (hunk 1 above)
Comment (enoodle) [reply docs/serving/autoscaling/autoscaling-metrics.md]: :+1:

File: docs/serving/autoscaling/autoscaling-metrics.md [hunk 3]
```diff
@@ -15,52 +15,89 @@ The possible metric types that can be configured per revision depend on the type
 For more information about KPA and HPA, see the documentation on [Supported Autoscaler types](autoscaler-types.md).
 
 * **Per-revision annotation key:** `autoscaling.knative.dev/metric`
-* **Possible values:** `"concurrency"`, `"rps"` or `"cpu"`, depending on your Autoscaler type. The `cpu` metric is only supported on revisions with the HPA class.
+* **Possible values:** `"concurrency"`, `"rps"`, `"cpu"`, `"memory"` or any custom metric name, depending on your Autoscaler type. The `cpu`/`memory`/`custom` metrics are only supported on revisions with the HPA class.
```
Comment (abrennan89) [inline docs/serving/autoscaling/autoscaling-metrics.md:18]:
```suggestion
* **Possible values:** `"concurrency"`, `"rps"`, `"cpu"`, `"memory"` or any custom metric name, depending on your Autoscaler type. The `"cpu"`, `"memory"`, and `"custom"` metrics are only supported on revisions that use the HPA class.
```

File: docs/serving/autoscaling/autoscaling-metrics.md [hunk 4]
```diff
@@ -15,52 +15,89 @@ The possible metric types that can be configured per revision depend on the type
...
 * **Default:** `"concurrency"`
 
 
-=== "Per-revision concurrency configuration"
-    ```yaml
-    apiVersion: serving.knative.dev/v1
-    kind: Service
+=== Per-revision concurrency configuration
```
Comment (abrennan89) [inline docs/serving/autoscaling/autoscaling-metrics.md:22]: @snneji can you confirm whether it was formatted like this before for tabs or something? I can't find it documented in the style guide.

File: docs/serving/autoscaling/autoscaling-metrics.md (hunk 3 above)
Comment (enoodle) [reply docs/serving/autoscaling/autoscaling-metrics.md]: :+1:

File: docs/serving/autoscaling/autoscaling-metrics.md (hunk 4 above)
Comment (snneji) [reply docs/serving/autoscaling/autoscaling-metrics.md:22]: Yes this used to be formatted as tabs. The 4 spaces of indentation these code snippets used to have needs to added back for the formatting to render correctly.

**Review by snneji — CHANGES_REQUESTED** [review]
Added correct formatting for 2 out of 4 code snippets to as suggested changes. Also made a tweak to wording.

File: docs/serving/autoscaling/autoscaling-metrics.md [hunk 5]
```diff
@@ -15,52 +15,89 @@ The possible metric types that can be configured per revision depend on the type
...
+metadata:
+  name: helloworld-go
+  namespace: default
+spec:
+  template:
+    metadata:
+      annotations:
+        autoscaling.knative.dev/class: "hpa.autoscaling.knative.dev"
+        autoscaling.knative.dev/metric: "metricName"
+```
```
[the remark this answers is not in the document]
Comment (enoodle) [reply docs/serving/autoscaling/autoscaling-metrics.md:100]: :+1:

File: docs/serving/autoscaling/autoscaling-metrics.md [hunk 6]
```diff
@@ -15,25 +15,58 @@ The possible metric types that can be configured per revision depend on the type
...
+
+```yaml
+apiVersion: serving.knative.dev/v1
+kind: Service
+metadata:
+  name: helloworld-go
+  namespace: default
+spec:
+  template:
     metadata:
```
Comment (snneji) [inline docs/serving/autoscaling/autoscaling-metrics.md:32]:
```suggestion
    ```yaml
    apiVersion: serving.knative.dev/v1
    kind: Service
    metadata:
      name: helloworld-go
      namespace: default
    spec:
      template:
        metadata:
```

File: docs/serving/autoscaling/autoscaling-metrics.md [hunk 7]
```diff
@@ -15,25 +15,58 @@ The possible metric types that can be configured per revision depend on the type
...
-      namespace: default
-    spec:
-      template:
-        metadata:
-          annotations:
-            autoscaling.knative.dev/metric: "concurrency"
-    ```
+      annotations:
+        autoscaling.knative.dev/metric: "concurrency"
+```
```
Comment (snneji) [inline docs/serving/autoscaling/autoscaling-metrics.md:35]:
```suggestion
          annotations:
            autoscaling.knative.dev/metric: "concurrency"
```
```

File: docs/serving/autoscaling/autoscaling-metrics.md [hunk 8]
```diff
@@ -15,25 +15,58 @@ The possible metric types that can be configured per revision depend on the type
...
+kind: Service
+metadata:
+  name: helloworld-go
+  namespace: default
+spec:
+  template:
+    metadata:
+      annotations:
+        autoscaling.knative.dev/metric: "rps"
+```
```
Comment (snneji) [inline docs/serving/autoscaling/autoscaling-metrics.md:50]:
```suggestion
    ```yaml
    apiVersion: serving.knative.dev/v1
    kind: Service
    metadata:
      name: helloworld-go
      namespace: default
    spec:
      template:
        metadata:
          annotations:
            autoscaling.knative.dev/metric: "rps"
    ```
```

File: docs/serving/autoscaling/autoscaling-metrics.md [hunk 9]
```diff
@@ -15,25 +15,58 @@ The possible metric types that can be configured per revision depend on the type
...
+metadata:
+  name: helloworld-go
+  namespace: default
+spec:
+  template:
+    metadata:
+      annotations:
+        autoscaling.knative.dev/class: "hpa.autoscaling.knative.dev"
+        autoscaling.knative.dev/metric: "cpu"
+```
```
Comment (snneji) [inline docs/serving/autoscaling/autoscaling-metrics.md:66]:
```suggestion
    ```yaml
    apiVersion: serving.knative.dev/v1
    kind: Service
    metadata:
      name: helloworld-go
      namespace: default
    spec:
      template:
        metadata:
          annotations:
            autoscaling.knative.dev/class: "hpa.autoscaling.knative.dev"
            autoscaling.knative.dev/metric: "cpu"
    ```
```

File: docs/serving/autoscaling/autoscaling-metrics.md [hunk 10]
```diff
@@ -15,25 +15,58 @@ The possible metric types that can be configured per revision depend on the type
...
-* **Possible values:** `"concurrency"`, `"rps"` or `"cpu"`, depending on your Autoscaler type. The `cpu` metric is only supported on revisions with the HPA class.
+* **Possible values:** `"concurrency"`, `"rps"`, `"cpu"`, `"memory"` or any custom metric name, depending on your Autoscaler type. The `"cpu"`, `"memory"`, and `"custom"` metrics are only supported on revisions that use the HPA class.
 * **Default:** `"concurrency"`
 
 
-=== "Per-revision concurrency configuration"
-    ```yaml
-    apiVersion: serving.knative.dev/v1
-    kind: Service
+=== Per-revision concurrency configuration
```
Comment (snneji) [inline docs/serving/autoscaling/autoscaling-metrics.md:22]:
```suggestion
=== "Per-revision concurrency configuration"
```

File: docs/serving/autoscaling/autoscaling-metrics.md [hunk 11]
```diff
@@ -15,25 +15,58 @@ The possible metric types that can be configured per revision depend on the type
...
-      template:
-        metadata:
-          annotations:
-            autoscaling.knative.dev/metric: "concurrency"
-    ```
+      annotations:
+        autoscaling.knative.dev/metric: "concurrency"
+```
+
+=== Per-revision rps configuration
```
Comment (snneji) [inline docs/serving/autoscaling/autoscaling-metrics.md:37]:
```suggestion
=== "Per-revision rps configuration"
```

File: docs/serving/autoscaling/autoscaling-metrics.md [hunk 12]
```diff
@@ -15,25 +15,58 @@ The possible metric types that can be configured per revision depend on the type
...
+  name: helloworld-go
+  namespace: default
+spec:
+  template:
+    metadata:
+      annotations:
+        autoscaling.knative.dev/metric: "rps"
+```
+
+=== Per-revision cpu configuration
```
Comment (snneji) [inline docs/serving/autoscaling/autoscaling-metrics.md:52]:
```suggestion
=== "Per-revision cpu configuration"
```

File: docs/serving/autoscaling/autoscaling-metrics.md [hunk 13]
```diff
@@ -15,25 +15,58 @@ The possible metric types that can be configured per revision depend on the type
...
+  namespace: default
+spec:
+  template:
+    metadata:
+      annotations:
+        autoscaling.knative.dev/class: "hpa.autoscaling.knative.dev"
+        autoscaling.knative.dev/metric: "cpu"
+```
+
+=== Per-revision memory configuration
```
Comment (snneji) [inline docs/serving/autoscaling/autoscaling-metrics.md:68]:
```suggestion
=== "Per-revision memory configuration"
```

File: docs/serving/autoscaling/autoscaling-metrics.md [hunk 14]
```diff
@@ -44,10 +77,15 @@ For more information about KPA and HPA, see the documentation on [Supported Auto
       template:
         metadata:
           annotations:
-            autoscaling.knative.dev/metric: "rps"
+            autoscaling.knative.dev/class: "hpa.autoscaling.knative.dev"
+            autoscaling.knative.dev/metric: "memory"
     ```
 
-=== "Per-revision cpu configuration"
+=== Per-revision custom metric configuration
```
Comment (snneji) [inline docs/serving/autoscaling/autoscaling-metrics.md:84]:
```suggestion
=== "Per-revision custom metric configuration"
```

**Review by snneji — CHANGES_REQUESTED** [review]
A few more things to fix the tab formatting.

**Review by snneji — APPROVED** [review]

**Comment by nader-ziada** [comment]
/hold cancel
/lgtm
pr in serving merged today

**Pull request merged**